packages/berry-core/sources/Configuration.ts

```diff
@@ -257,24 +272,85 @@ function parseBoolean(value: unknown) {
+  if (Array.isArray(value)) {
+    if (!definition.isArray && !Array.isArray(definition.default)) {
+      throw new Error(`Non-array configuration settings "${path}" cannot be an array`);
+    } else {
+      return value.map((sub, i) => parseSingleValue(`${path}[${i}]`, sub, definition, folder));
+    }
+  } else {
+    return parseSingleValue(path, value, definition, folder);
+  }
+}
+
+function parseSingleValue(path: string, value: unknown, definition: SettingsDefinition, folder: string) {
+  switch(definition.type) {
+    case SettingsType.BOOLEAN:
+      return parseBoolean(value);
+    case SettingsType.OBJECT:
+      return parseObject(path, value, definition, folder);
+      case SettingsType.MAP:
+      return parseMap(path, value, definition, folder);
+  }
 
   if (typeof value !== `string`)
     throw new Error(`Expected value to be a string`);
 
-  if (type === SettingsType.ABSOLUTE_PATH) {
-    return posix.resolve(folder, NodeFS.toPortablePath(value));
-  } else if (type === SettingsType.LOCATOR_LOOSE) {
-    return structUtils.parseLocator(value, false);
-  } else if (type === SettingsType.LOCATOR) {
-    return structUtils.parseLocator(value);
-  } else {
-    return value;
+  switch(definition.type) {
+    case SettingsType.ABSOLUTE_PATH:
+      return posix.resolve(folder, NodeFS.toPortablePath(value));
+    case SettingsType.LOCATOR_LOOSE:
+      return structUtils.parseLocator(value, false);
+    case SettingsType.LOCATOR:
+      return structUtils.parseLocator(value);
+    default:
+      return value;
   }
 }
 
+function parseObject(path: string, value: unknown, definition: ObjectSettingsDefinition, folder: string) {
+  const result = new Map<string, any>();
+
+  if (typeof value !== `object` || Array.isArray(value))
+    throw new Error(`Object configuration settings "${path}" must be an object`);
+
+  for (const [propKey, propValue] of Object.entries(value!)) {
+    const name = normalizeKey(propKey);
+    const subPath = `${path}.${name}`;
+    const subDefinition = definition.properties[name];
+
+    if (!subDefinition)
+      throw new UsageError(`Unrecognized configuration settings found: ${path}.${propKey} - run "yarn config -v" to see the list of settings supported in Yarn`);
+
+    result.set(name, parseValue(subPath, propValue, definition.properties[name], folder));
+  }
+
+  return result;
+}
+
+function parseMap(path: string, value: unknown, definition: MapSettingsDefinition, folder: string) {
+  const result = new Map<string, any>();
+
+  if (typeof value !== 'object' || Array.isArray(value))
+    throw new Error(`Map configuration settings "${path}" must be an object`);
+
+  for (const [propKey, propValue] of Object.entries(value!)) {
+    const subPath = `${path}['${propKey}']`;
+
+    result.set(propKey, parseValue(subPath, propValue, definition.valueDefinition, folder));
+  }
+
+  return result;
+}
+
+function normalizeKey(key: string) {
+  return key.replace(/[_-]([a-z])/g, ($0, $1) => $1.toUpperCase());
```

Comment:
I think since then I've added `camelCase` to the core - we probably should use it instead of this one-liner